doc/neps/nep-XXXX-SIMD-optimizations.rst

```diff
@@ -0,0 +1,249 @@
+Single Instruction Multiple Data (SIMD) constructs. This NEP proposes a similar
+mechanism for NumPy. There are three stages to using the mechanism:
+- Infrastructure is provided in the code for abstract intrinsics. The ufunc
+  machinery will be extended using sets of these abstract intrinsics, so that
+  a single ufunc will be expressed as a set of loops, going from a minimal to
+  a maximal set of possibly availabe intrinsics.
+- At compile time, compiler macros and CPU detection are used to turn the
+  abstract intrinsics into concrete intrinsic calls. Any intrinsics not
+  available on the platform, either because the CPU does not support them
+  (and so cannot be tested) or because the abstract intrinsic does not have a
+  parallel concrete intrinsic on the platform will not error, rather the
+  corresponding loop will not be produced and added to the set of
+  possibilities.
+- At runtime, the CPU detection code will further limit the set of loops
+  available, and the optimal one will be chosen for the ufunc.
+
+The current NEP proposes only to use the runtime feature detection and optimal
+loop selection mechanism for ufuncs. Future NEPS may propose other uses for the
+proposed solution.
+
+Usage and Impact
+----------------
+
+The end user will be able to get a list of intrinsics available for their
+platform and compiler. Optionally,
+the user may be able to specify which of the loops available at runtime will be
+used, perhaps via an environment variable to enable benchmarking the impact of
+the different loops. There should be no direct impact to naive end users, the
+results of all the loops should be identical to within a small number (1-3)
+ULPs. On the other hand, users with more powerful machines should notice a
+performance boost.
+
+Binary releases - wheels on PyPI and conda packages
+```````````````````````````````````````````````````
+
+The binaries released by this process will be larger since they include all
+possible loops for the architecture. Some packagers may prefer to limit the
+number of loops in order to limit the size of the binaries, we would hope they
+would still support a wide range of families of architectures. Note this
+problem already exists in the Intel MKL offering.
+
+Source builds
+`````````````
+TBD
+- Setting the baseline and set of runtime-dispatchable ISA extensions
+- Behavior when compiler or hardware doesn't support a requested ISA extension
+
+
+How to run benchmarks to assess performance benefits
+````````````````````````````````````````````````````
+
+Adding more code which use intrinsics will make the code harder to maintain.
+Therefore, such code should only be added if it yields a significant
+performance benefit. Assessing this performance benefit can be nontrivial.
+To aid with this, the implementation for this NEP will add a way to select
+which instruction sets can be used at *runtime* via environment variables.
+(name TBD).
```

Comment:
I think a maybe even more important thing is that we must be able to run all/all the code branches in CI to ensure that the results are identical? Tools such as code coverage may also be tricky.

Reply:
adding. Note that CI reporting of code coverage seems to be broken.